In [ ]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [5]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()

# Fit and transform the data using StandardScaler
df_market_data_normalized = pd.DataFrame(scaler.fit_transform(df_market_data), columns=df_market_data.columns, index=df_market_data.index)

# Display sample data after normalization
print("\nSample data after normalization:")
df_market_data_normalized.head(10)


Sample data after normalization:


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                           -0.251637  
ethereum                          -0.199352  
tether                            -0.282061  
ripple                            -0.295546  
bitcoin-cash                      -0.270317  
binancecoin                       -0.225533  
chainlink                         -0.018284  
cardano                           -0.155428  
litecoin                          -0.292351  
bitcoin-cash-sv                   -0.206029

---

### Find the Best Value for k Using the Original Data.

In [7]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 11))

In [8]:
import warnings
warnings.filterwarnings("ignore")

In [9]:
# Create an empty list to store the inertia values
inertia_values = []
# Create a for loop to compute the inertia with each possible value of k

for i in k_values:
    kmeans = KMeans(n_clusters=i, random_state=1)
    kmeans.fit(df_market_data_normalized)
    inertia_values.append(kmeans.inertia_)


In [10]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    'Number of Clusters (k)': k_values,
    'Inertia': inertia_values
}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)
df_elbow

Number of Clusters (k)     Inertia
0                       1  287.000000
1                       2  195.820218
2                       3  123.190482
3                       4   79.022435
4                       5   65.405923
5                       6   52.933559
6                       7   47.983124
7                       8   37.288187
8                       9   33.061685
9                      10   28.779752

In [11]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_plot = df_elbow.hvplot.line(x='Number of Clusters (k)', 
                                  y='Inertia', 
                                  title='Elbow Curve for Optimal k', 
                                  xlabel='Number of Clusters (k)', 
                                  ylabel='Inertia', 
                                  grid=True)

# Display the interactive plot
elbow_plot

:Curve   [Number of Clusters (k)]   (Inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:**  **4**


---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [12]:
# Initialize the K-Means model using the best value for k
best_k = 4
kmeans_model = KMeans(n_clusters=best_k, random_state=1)

In [13]:
# Fit the K-Means model using the scaled data
kmeans_model.fit(df_market_data_normalized)

KMeans(n_clusters=4, random_state=1)

In [14]:
# Predict the clusters to group the cryptocurrencies using the scaled data
crypto_clusters = kmeans_model.predict(df_market_data_normalized)

# Print the resulting array of cluster values
print(crypto_clusters)

[3 3 1 1 3 3 3 3 3 1 1 1 1 3 1 3 1 1 3 1 1 3 1 1 1 1 1 1 3 1 1 1 0 3 1 1 2
 1 1 1 1]


In [40]:
# Create a copy of the DataFrame
with_crypto_clusters = df_market_data_normalized.copy()

In [47]:
with_crypto_clusters['cluster_label'] = crypto_clusters

# Display sample data
with_crypto_clusters.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  cluster_label  
coin_id                                                     
bitcoin                           -0.251637              3  
ethereum                          -0.199352              3  
tether                            -0.282061              1  
ripple                            -0.295546              1  
bitcoin-cash                      -0.270317              3  
binancecoin                       -0.225533              3  
chainlink                         -0.018284              3  
cardano                           -0.155428              3  
litecoin                          -0.292351              3  
bitcoin-cash-sv                   -0.206029              1

In [48]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

# Create a scatter plot using hvPlot
scatter_plot = with_crypto_clusters.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="cluster_label",  # Use cluster_label column for coloring
    hover_cols=["coin_id"],  # Add the crypto name for hover information
    xlabel="Price Change Percentage (24h)",
    ylabel="Price Change Percentage (7d)",
    title="Cryptocurrency Clusters"
)

# Display the scatter plot
scatter_plot


:NdOverlay   [cluster_label]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [18]:
# Create a PCA model instance and set `n_components=3`.
pca_model = PCA(n_components=3)

In [62]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
# Fit the PCA model to the scaled data
pca_model.fit(df_market_data_normalized)

#Transform the data to the new coordinate system (3 principal components)
df_pca_transformed = pca_model.transform(df_market_data_normalized)

# Create a DataFrame with the transformed data
df_pca_components = pd.DataFrame(df_pca_transformed, columns=['PCA_Component_1', 'PCA_Component_2', 'PCA_Component_3'], index=df_market_data_normalized.index)

# View the first five rows of the DataFrame. 
df_pca_components

PCA_Component_1  PCA_Component_2  PCA_Component_3
coin_id                                                                
bitcoin                     -0.600667         0.842760         0.461595
ethereum                    -0.458261         0.458466         0.952877
tether                      -0.433070        -0.168126        -0.641752
ripple                      -0.471835        -0.222660        -0.479053
bitcoin-cash                -1.157800         2.041209         1.859715
binancecoin                 -0.516534         1.388377         0.804071
chainlink                   -0.450711         0.517699         2.846143
cardano                     -0.345600         0.729439         1.478013
litecoin                    -0.649468         0.432165         0.600303
bitcoin-cash-sv             -0.759014        -0.201200        -0.217653
crypto-com-chain            -0.248198        -1.376252        -1.462026
usd-coin                    -0.438408        -0.175337        -0.663388
eos                         -0.693425        -0.473815        -0.527597
monero                       0.060499         2.909404         1.498571
tron                        -0.393352        -0.108192        -0.012756
tezos                       -0.796176        -0.494409         1.082812
okb                          0.064075        -1.269825        -1.098829
stellar                     -0.489015        -0.732719        -0.062543
cosmos                      -0.306272         0.703415         1.714224
cdai                        -0.513528        -0.142802        -0.656566
neo                         -0.362120        -0.986914        -0.728752
wrapped-bitcoin             -0.604265         0.827398         0.439316
leo-token                   -0.413296        -0.674115        -1.076628
huobi-token                 -0.407483        -0.212507        -0.351426
nem                          0.608974         0.563532        -1.148742
binance-usd                 -0.450211        -0.151019        -0.647401
iota                        -0.764665        -0.517886         0.204990
vechain                     -0.556315        -1.938209        -1.261776
zcash                       -0.425147         0.492976         1.058048
theta-token                  2.676868        -0.013954        -1.965207
dash                        -0.613923        -0.479337         0.339565
ethereum-classic            -0.579924        -0.356334        -0.114942
ethlend                      8.089018        -3.896891         2.301382
maker                       -0.389045         0.165041         0.379414
havven                       0.865762        -2.261882         0.275583
omisego                      0.111675         0.428316        -1.205398
celsius-degree-token         4.792395         6.767679        -1.986985
ontology                    -0.632355        -2.108117        -0.652227
ftx-token                   -0.593142         0.021485         0.209911
true-usd                    -0.458131        -0.135734        -0.635284
digibyte                    -0.297910        -0.191126        -0.909602

In [20]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
# Retrieve the explained variance
explained_variance = pca_model.explained_variance_ratio_

# Display the explained variance for each principal component
for i, variance in enumerate(explained_variance, 1):
    print(f"Explained Variance for PCA Component {i}: {variance:.4f}")

total_explained_variance = sum(explained_variance)
print(f"Total Explained Variance of the Three Principal Components: {total_explained_variance:.4f}")

Explained Variance for PCA Component 1: 0.3720
Explained Variance for PCA Component 2: 0.3470
Explained Variance for PCA Component 3: 0.1760
Total Explained Variance of the Three Principal Components: 0.8950


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.8950

In [21]:
# Create a new DataFrame with the PCA data.

# Creating a DataFrame with the PCA data

# Copy the crypto names from the original data


# Set the coinid column as index


# Display sample data


---

### Find the Best Value for k Using the PCA Data

In [36]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 11))

In [37]:
# Create an empty list to store the inertia values
inertia_values = []
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k_values:
    kmeans = KMeans(n_clusters=i, random_state=1)
    kmeans.fit(df_pca_components)
    inertia_values.append(kmeans.inertia_)

In [38]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_2 = {
    'Number of Clusters (k)': k_values,
    'Inertia': inertia_values
}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_2 = pd.DataFrame(elbow_data_2)
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_2

Number of Clusters (k)     Inertia
0                       1  256.874086
1                       2  165.901994
2                       3   93.774626
3                       4   49.665497
4                       5   38.352251
5                       6   27.618972
6                       7   21.134056
7                       8   17.437664
8                       9   13.742792
9                      10   10.484890

In [61]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_plot_2 = df_elbow_2.hvplot.line(x='Number of Clusters (k)', 
                                  y='Inertia', 
                                  title='Elbow Curve for Optimal k with PCA', 
                                  xlabel='Number of Clusters (k)', 
                                  ylabel='Inertia', 
                                  grid=True)

# Display the interactive plot
elbow_plot_2

:Curve   [Number of Clusters (k)]   (Inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** NO

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [44]:
# Initialize the K-Means model using the best value for k
best_k = 4
kmeans_model = KMeans(n_clusters=best_k, random_state=1)

In [45]:
# Fit the K-Means model using the PCA data
# Fit the K-Means model using the scaled data
kmeans_model.fit(df_pca_components)

KMeans(n_clusters=4, random_state=1)

In [46]:
# Predict the clusters to group the cryptocurrencies using the PCA data
crypto_clusters_2 = kmeans_model.predict(df_pca_components)

# Print the resulting array of cluster values
print(crypto_clusters_2)


[1 1 0 0 1 1 1 1 1 0 0 0 0 1 0 1 0 0 1 0 0 1 0 0 0 0 0 0 1 0 0 0 2 1 0 0 3
 0 0 0 0]


In [54]:
# Create a copy of the DataFrame with the PCA data
with_crypto_clusters_2 = df_pca_components.copy()

# Add a new column to the DataFrame with the predicted clusters
with_crypto_clusters_2['cluster_label'] = crypto_clusters_2

# Display sample data
with_crypto_clusters_2.head()


PCA_Component_1  PCA_Component_2  PCA_Component_3  cluster_label
coin_id                                                                       
bitcoin             -0.600667         0.842760         0.461595              1
ethereum            -0.458261         0.458466         0.952877              1
tether              -0.433070        -0.168126        -0.641752              0
ripple              -0.471835        -0.222660        -0.479053              0
bitcoin-cash        -1.157800         2.041209         1.859715              1

In [55]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
# Create a scatter plot using hvPlot
scatter_plot_2 = with_crypto_clusters_2.hvplot.scatter(
    x="PCA_Component_1",
    y="PCA_Component_2",
    by="cluster_label",  # Use cluster_label column for coloring
    hover_cols=["coin_id"],  # Add the crypto name for hover information
    xlabel="PC1",
    ylabel="PC2",
    title="Cryptocurrency Clusters with PCA"
)

# Display the scatter plot
scatter_plot_2

:NdOverlay   [cluster_label]
   :Scatter   [PCA_Component_1]   (PCA_Component_2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [58]:
# Composite plot to contrast the Elbow curves
composite_plot_elbow = elbow_plot + elbow_plot_2

# Display the composite plot
composite_plot_elbow

:Layout
   .Curve.I  :Curve   [Number of Clusters (k)]   (Inertia)
   .Curve.II :Curve   [Number of Clusters (k)]   (Inertia)

In [60]:
# Composite plot to contrast the clusters
composite_plot_clusters = scatter_plot + scatter_plot_2

# Display the composite plot
composite_plot_clusters

:Layout
   .NdOverlay.I  :NdOverlay   [cluster_label]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [cluster_label]
      :Scatter   [PCA_Component_1]   (PCA_Component_2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The PCA Data Looks better a little. While the clusters are a little better defined, once you look at the graph when it shared the same scale, it is not that much different. However, the outliers are more easily decernable and look more like outliers.